### ISSUE TITLE
Support AbortSignal for aborting requests

### ISSUE THREAD

**Issue by benjamingr** [axios/axios#3506]
#### Is your feature request related to a problem? Please describe.

Node 15 has built-in support for `AbortController` and `AbortSignal` and request cancellation propagates all the way up to [`http.request`](https://nodejs.org/api/http.html#http_http_request_url_options_callback) taking a signal parameter.

It would be useful to be able to abort an axios request using `AbortSignal`s:

```js
const ac = new AbortController();
const { signal } = ac;

axios({
  method: 'get',
  url: 'http://example.com',
  signal,
}).then(...);
ac.abort(); // aborts the above request with an `AbortError`
```

#### Describe the solution you'd like

Basically, axios already supports `CancelToken`s which are very similar to `AbortSignal`s.

Basically:

 - Rename `cancelToken` to `signal` (and support the old API for compatibility.
 - Allow passing an `AbortSignal` in. 
 - Listen to `abort` instead of doing `.promise.then` on the cancel token.

#### Describe alternatives you've considered

Since AbortSignal is the primitive both the web and Node.js settled on - alternatives that I'd prefer would likely need to go through those bodies first probably.

#### Additional context

I'm happy to help with any issues or missing APIs from the Node.js side.

**Comment by benjamingr** [comment]
Basically replacing this code

https://github.com/axios/axios/blob/e426910be7c417bdbcde9c18cb184ead826fc0e1/lib/adapters/http.js#L284-L292

With no code at all since `http.request` already takes a signal which will take care of aborting the request with the appropriate error for you.

**Comment by DigitalBrainJS** [comment]
#3305 :)
